kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-common/src/main/java/org/kie/server/remote/rest/common/util/WrapperMessageBodyWriter.java

```diff
@@ -0,0 +1,59 @@
+package org.kie.server.remote.rest.common.util;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Type;
+import java.util.Map;
+
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.MultivaluedMap;
+import javax.ws.rs.ext.MessageBodyWriter;
+import javax.ws.rs.ext.Provider;
+
+import org.kie.server.api.KieServerConstants;
+import org.kie.server.api.marshalling.Marshaller;
+import org.kie.server.api.marshalling.MarshallerFactory;
+import org.kie.server.api.marshalling.MarshallingFormat;
+import org.kie.server.remote.rest.common.util.RestUtils.Wrapper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@Provider
+public class WrapperMessageBodyWriter implements MessageBodyWriter<Wrapper<?>> {
+
+    private static final Logger logger = LoggerFactory.getLogger(WrapperMessageBodyWriter.class);
+
+    private static final boolean STRICT_ID_JSON = Boolean.parseBoolean(System.getProperty(KieServerConstants.KIE_SERVER_STRICT_ID_JSON, "false"));
+
+
+
+    @Override
+    public boolean isWriteable(Class<?> type, Type genericType, Annotation[] annotations, MediaType mediaType) {
+        return type == Wrapper.class;
+    }
+
+    @Override
+    public void writeTo(Wrapper<?> identifier,
+                        Class<?> type,
+                        Type genericType,
+                        Annotation[] annotations,
+                        MediaType mediaType,
+                        MultivaluedMap<String, Object> httpHeaders,
+                        OutputStream entityStream) throws IOException, WebApplicationException {
+
+        Map<String, String> parameters = mediaType.getParameters();
+        Boolean isStrict = Boolean.parseBoolean(parameters.get("strict"));
+
+        MarshallingFormat format = MarshallingFormat.fromType(mediaType.toString());
+        Marshaller marshaller = MarshallerFactory.getMarshaller(format, Thread.currentThread().getContextClassLoader());
+        if ((isStrict || STRICT_ID_JSON) && MediaType.APPLICATION_JSON_TYPE.isCompatible(mediaType)) {
+            entityStream.write(marshaller.marshall(identifier).getBytes());
+        } else {
+            entityStream.write(identifier.getId().toString().getBytes());
```

Comment:
This is one way how to do it.
The other thing I was thinking about is to just simply create a wrapper JAXB object, something like
https://github.com/kiegroup/droolsjbpm-integration/blob/7d93d03f9de8a68d78d04ba25d71bdc80eaa3112/kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/model/definition/ProcessDefinitionList.java#L30-L63

And then we can decide for example directly in a JAX-RS endpoint method whether we want to wrap a returned number inside that wrapper or not based on whether the `strict` is set either in a header or globally.

If it is not set at all, we won't wrap e.g. the process instance id and will return just text.
If it is set, then we will wrap it inside the annotated wrapper and either let JAX-RS take care of it automatically or perhaps use our own marshaller and marshall it.

I think this approach requires less code, works for both XML and JSON and doesn't require a custom BodyWriter. What do you think?

Reply:
Hi Marian, Regarding the spec I argue that your approach is not align.

4.2.2 Message Body Writer
The MessageBodyWriter interface defines the contract between the JAX-RS runtime and components that
provide mapping services from a Java type to a representation. A class wishing to provide such a service
implements the MessageBodyWriter interface and may be annotated with @Provider for automatic
discovery.

So the responsible for providing the representation of a java object is that interface. 

I did wrap already the object in an Wrapper and it is up to the MessageBodyWriter to decided how it is represented. Add that to the endpoint mixes responsabilities. this impl makes more sense to me as it is more align to that definition. In the end when you produce a mime type you are using default providers.

Reply:
Well, it was just an idea :) But then we mix responsibilities already with this ProcessDefinitionList object. We don't register any provider for that, we just marshal it ourselves. Also I don't think we use any default providers because before returning the response object we already marshal it using our internal marshaller into String. Which (probably) means that JAX-RS will leave it as is.
As I said, it was just an idea with less code. It would work the same also for XML whereas your approach needs to modify the Provider for XML as well. So it's up to you.